bin/merge_bins.py

```diff
@@ -0,0 +1,245 @@
+        for phylogeny in self.parse_tree(tree_qa):
+            if phylogeny[0] == self.bin_1.id and self.bin_1.phylo == []:
+                self.bin_1.phylo = phylogeny[3:]
+            elif phylogeny[0] == self.bin_2.id and self.bin_2.phylo == []:
+                self.bin_2.phylo = phylogeny[3:]
+
+    @staticmethod
+    def parse_profile(profile_txt):
+        with open(profile_txt, "r") as p:
+            p.readline()
+            header = p.readline()
+            p.readline()
+            for l in p:
+                sl = l.split()
+                if len(sl) < 2:
+                    continue
+                yield(sl[:3])  # bin_id, bin size, n mapped reads
+
+    @staticmethod
+    def parse_tree(tree_qa):
+        with open(tree_qa, "r") as t:
+            t.readline()
+            header = t.readline()
+            t.readline()
+            for l in t:
+                sl = l.split()
+                if len(sl) < 2:
+                    continue
+                yield(sum([i.split(";") for i in sl], []))
+
+
+def get_mean(length_file):
+    with open(length_file, "r") as f:
+        total = [int(line.split()[0]) for line in f]
+    mean = sum(total) / len(total)
+    return(mean)
+
+
+def parse_merge_file(merger_tsv):
+    with open(merger_tsv, "r") as m:
+        m.readline()  # discard the header
+        for l in m:
+            line = l.split()
+            pm = PossibleMerge(line)
+            yield(pm)
+
+
+def concatenate(file_list, output):
+    """Concatenate files together
+    Args:
+        file_list (list): the list of input files (can be a generator)
+        output (string): the output file name
+    """
+    try:
+        out_file = open(output, "wb")
+    except (IOError, OSError) as e:
+        print("Failed to open output file: %s" % e)
+        sys.exit(1)
+
+    with out_file:
+        for file_name in file_list:
+            if file_name is not None:
+                with open(file_name, "rb") as f:
+                    copyfile(f, out_file)
+
+
+def merge(args):
+    all_bins = os.listdir(args.bins)
+    n_merged = 0
+
+    # merge bins according to criteria in
+    # https://doi.org/10.1038/s41564-017-0012-7
+    mean_read_length = get_mean(args.length)
+    for pm in parse_merge_file(args.merger):
+        pm.add_profile_info(args.profile, mean_read_length)
+        pm.add_tree_info(args.tree)
+        abs_delta_cov = \
+            max(pm.bin_1.read_cov, pm.bin_2.read_cov) /\
+            min(pm.bin_1.read_cov, pm.bin_2.read_cov)
+        clades = []
+        for p1, p2 in zip(pm.bin_1.phylo, pm.bin_2.phylo):
+            clades.append((p1, p2))
+
+        if pm.delta_contam <= args.delta_cont and \
+           pm.merged_contam <= args.merged_cont and \
+           pm.delta_compl >= args.delta_compl and \
+           abs_delta_cov >= args.abs_delta_cov and \
+           clades[-1][0] == clades[-1][1]:
+
+            # then read the two bins
+            bin_1_p = "%s/%s.fa" % (args.bins, pm.bin_1.id)
```

Comment:
This won't work if extension is `.fasta`... Perhaps it would be worth it reading in all bins into a dictionary with bin_id as key and filepath as values?